packages/runner/src/traverse.ts

```diff
@@ -864,158 +1398,563 @@ export class SchemaObjectTraverser<S extends BaseMemoryAddress>
+          curSelector.schemaContext!.rootSchema,
+        );
+        // Replace doc with a DataCellURI style doc
+        // TODO(@ubik2): ideally, we wouldn't use this path in query traversal.
+        // Right now, we aren't passing both the link info and doc info, so we
+        // will override the doc here.
+        // I could switch based off the traverseCells flag (true for queries),
+        // but I don't want to have that change behavior here.
+        curDoc = {
+          ...curDoc,
+          address: {
+            ...curDoc.address,
+            id: createDataCellURI(curDoc.value, elementLink),
+            path: ["value"],
+          },
+        };
+        // Our selector's path needs to be updated to match the new doc
+        curSelector.path = curDoc.address.path;
+      }
+      // If we've asked for cells in the array and we don't need to traverse cells,
+      // add the created cell instead.
+      if (
+        !this.traverseCells &&
+        SchemaObjectTraverser.asCellOrStream(itemSchema) &&
+        isPrimitiveCellLink(item)
+      ) {
+        // This is redundant, since the getAtPath that happened earlier
+        // already did this work, but I don't have easy access here.
+        const itemLink = parseLink(item, doc.address); // should trim path, but it doesn't matter
+        // The link may not have the asCell flags, so pull that from itemSchema
+        const elementLink = {
+          ...itemLink,
+          schema: mergeSchemaFlags(itemSchema, itemLink.schema ?? {}),
+        };
+        const val = this.objectCreator.createObject(elementLink, undefined);
+        arrayObj.push(val);
+      } else {
+        // FIXME: If we have an asCell in our schema,
+        // We want those links to point directly at the linked cells, instead of
+        // using our path (e.g. ["items", "0"]), so don't pass in a modified link.
+        const val = this.traverseWithSelector(curDoc, curSelector);
+        if (val === undefined) {
+          // this array is invalid, since one or more items do not match the schema
+          logger.debug(
+            "traverse",
+            () => ["Item doesn't match array schema", curDoc, curSelector],
+          );
+          return undefined;
+        }
+        arrayObj.push(val);
       }
-      arrayObj.push(val);
     }
     return arrayObj;
   }
 
+  /**
+   * Traverse an object according to the specified schema, returning
+   * a new object that only includes the properties that matched the schema.
+   *
+   * When properties are specified in the schema, and additionalProperties
+   * is not specified, properties not in the schema are not traversed, but
+   * the objectCreator can control whether these optional properties are
+   * included.
+   *
+   * When properties are not specified in the schema, and additionalProperties
+   * is not specified, we have the standard JSONSchema behavior, where this is
+   * equivalent to additionalProperties of true.
```

Comment:
do we actually see this at this point or would have this been collapsed to a true schema earlier?